src/rust/src/builder_ffi.rs

```diff
@@ -0,0 +1,169 @@
+    bundle::{Authorized, Flags},
+    keys::OutgoingViewingKey,
+    value::NoteValue,
+    Bundle,
+};
+use rand_core::OsRng;
+use tracing::error;
+use zcash_primitives::transaction::{
+    components::{sapling, transparent, Amount},
+    sighash::{SignableInput, SIGHASH_ALL},
+    sighash_v5::v5_signature_hash,
+    Authorization, TransactionData, TxVersion,
+};
+
+use crate::{transaction_ffi::PrecomputedTxParts, ORCHARD_PK};
+
+#[no_mangle]
+pub extern "C" fn orchard_builder_new(
+    spends_enabled: bool,
+    outputs_enabled: bool,
+    anchor: *const [u8; 32],
+) -> *mut Builder {
+    let anchor = unsafe { anchor.as_ref() }.expect("Anchor pointer may not be null.");
+    Box::into_raw(Box::new(Builder::new(
+        Flags::from_parts(spends_enabled, outputs_enabled),
+        orchard::Anchor::from_bytes(*anchor).unwrap(),
+    )))
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_builder_add_recipient(
+    builder: *mut Builder,
+    ovk: *const [u8; 32],
+    recipient: *const orchard::Address,
+    value: u64,
+    memo: *const [u8; 512],
+) -> bool {
+    let builder = unsafe { builder.as_mut() }.expect("Builder may not be null.");
+    let ovk = unsafe { ovk.as_ref() }
+        .copied()
+        .map(OutgoingViewingKey::from);
+    let recipient = unsafe { recipient.as_ref() }.expect("Recipient may not be null.");
+    let value = NoteValue::from_raw(value);
+    let memo = unsafe { memo.as_ref() }.expect("Memo may not be null.");
+
+    match builder.add_recipient(ovk, *recipient, value, Some(*memo)) {
+        Ok(()) => true,
+        Err(e) => {
+            error!("Failed to add Orchard recipient: {}", e);
+            false
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_builder_free(builder: *mut Builder) {
+    if !builder.is_null() {
+        drop(unsafe { Box::from_raw(builder) });
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_builder_build(
+    builder: *mut Builder,
+) -> *mut Bundle<InProgress<Unproven, Unauthorized>, Amount> {
+    if builder.is_null() {
+        error!("Called with null builder");
+        return ptr::null_mut();
+    }
+    let builder = unsafe { Box::from_raw(builder) };
+
+    match builder.build(OsRng) {
+        Ok(bundle) => Box::into_raw(Box::new(bundle)),
+        Err(e) => {
+            error!("Failed to build Orchard bundle: {:?}", e);
+            ptr::null_mut()
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_unauthorized_bundle_free(
+    bundle: *mut Bundle<InProgress<Unproven, Unauthorized>, Amount>,
+) {
+    if !bundle.is_null() {
+        drop(unsafe { Box::from_raw(bundle) });
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_unauthorized_bundle_prove_and_sign(
+    bundle: *mut Bundle<InProgress<Unproven, Unauthorized>, Amount>,
+    sighash: *const [u8; 32],
+) -> *mut Bundle<Authorized, Amount> {
+    let bundle = unsafe { Box::from_raw(bundle) };
+    let sighash = unsafe { sighash.as_ref() }.expect("sighash pointer may not be null.");
```

Comment:
Since `[u8; 32]` implements `Copy`, would `unsafe { &*sighash }` work here? What's the advantage one way or another? I ask because it seems to me like we aren't consistent about our idiom for passing fixed-length arrays across.

Reply:
Yes it would, and we in fact dereference (to copy) `sighash` below. I personally prefer to minimise the amount of trickiness that goes on inside `unsafe {}` blocks, so here I'm doing the minimal work to convert from `*const [u8; 32]` to `Option<&[u8; 32]>`, and then obtaining a `[u8; 32]` from that in the obvious Rust way for something we are assuming is not `None`.